zarr/v3/group.py

```diff
@@ -1,97 +1,106 @@
 
 
 @frozen
-class Group:
+class AsyncGroup(AsyncGroup):
     metadata: GroupMetadata
     store_path: StorePath
     runtime_configuration: RuntimeConfiguration
 
     @classmethod
-    async def create_async(
+    async def create(
         cls,
         store: StoreLike,
         *,
         attributes: Optional[Dict[str, Any]] = None,
         exists_ok: bool = False,
+        zarr_format: Literal[2, 3] = 3,  # field(default=3, validator=validators.in_([2, 3])),
         runtime_configuration: RuntimeConfiguration = RuntimeConfiguration(),
-    ) -> Group:
+    ) -> "AsyncGroup":
         store_path = make_store_path(store)
         if not exists_ok:
-            assert not await (store_path / ZARR_JSON).exists_async()
+            if zarr_format == 3:
+                assert not await (store_path / ZARR_JSON).exists_async()
+            elif zarr_format == 2:
+                assert not await (store_path / ZGROUP_JSON).exists_async()
+        print(zarr_format, type(zarr_format))
         group = cls(
-            metadata=GroupMetadata(attributes=attributes or {}),
+            metadata=GroupMetadata(attributes=attributes or {}, zarr_format=zarr_format),
             store_path=store_path,
             runtime_configuration=runtime_configuration,
         )
         await group._save_metadata()
         return group
 
     @classmethod
-    def create(
-        cls,
-        store: StoreLike,
-        *,
-        attributes: Optional[Dict[str, Any]] = None,
-        exists_ok: bool = False,
-        runtime_configuration: RuntimeConfiguration = RuntimeConfiguration(),
-    ) -> Group:
-        return sync(
-            cls.create_async(
-                store,
-                attributes=attributes,
-                exists_ok=exists_ok,
-                runtime_configuration=runtime_configuration,
-            ),
-            runtime_configuration.asyncio_loop,
-        )
-
-    @classmethod
-    async def open_async(
+    async def open(
         cls,
         store: StoreLike,
         runtime_configuration: RuntimeConfiguration = RuntimeConfiguration(),
-    ) -> Group:
+        zarr_format: Literal[2, 3] = 3,
+    ) -> "AsyncGroup":
         store_path = make_store_path(store)
-        zarr_json_bytes = await (store_path / ZARR_JSON).get_async()
-        assert zarr_json_bytes is not None
-        return cls.from_json(store_path, json.loads(zarr_json_bytes), runtime_configuration)
 
-    @classmethod
-    def open(
-        cls,
-        store: StoreLike,
-        runtime_configuration: RuntimeConfiguration = RuntimeConfiguration(),
-    ) -> Group:
-        return sync(
-            cls.open_async(store, runtime_configuration),
-            runtime_configuration.asyncio_loop,
-        )
+        # TODO: consider trying to autodiscover the zarr-format here
+        if zarr_format == 3:
+            # V3 groups are comprised of a zarr.json object
+            # (it is optional in the case of implicit groups)
```

Comment:
Are implicit groups really a thing? I don't think they show up in the v3 spec.

Reply:
They do show up in the spec (in quite a few places actually)! https://zarr-specs.readthedocs.io/en/latest/v3/core/v3.0.html#explicit-vs-implicit-groups